storage/innobase/trx/trx0rseg.cc

```diff
@@ -211,13 +207,11 @@ bool trx_rseg_read_wsrep_checkpoint(XID& xid)
 		trx_id_t id = mach_read_from_8(rseg_header
 					       + TRX_RSEG_MAX_TRX_ID);
 
-		if (id < max_id) {
-			continue;
+		if ((!found || id > max_id) &&
+		    trx_rseg_read_wsrep_checkpoint(rseg_header, xid)) {
+			max_id = id;
+			found = true;
```

Comment:
If write-set replication is disabled after the database server has run for quite some time, then it could happen that a WSREP XID is already present in all rollback segment header pages. In that case, we would still seem to return a possibly older WSREP XID.

Let’s assume that rollback segment header page A contained the latest WSREP XID before `SET wsrep_on=0` was executed. Now, let us assume that after that, the latest commit is on rollback segment header page B, which also happened to contain some older WSREP XID.

It seems to me that we should choose the maximum WSREP XID in this case, instead of choosing the WSREP XID of the rollback segment header page with the latest commit.

For this, we would need something that can compare (sort) WSREP XIDs. Can you explain how it would work?

Similarly, we should choose the maximum binary log file and position, instead of choosing the one on the rollback segment header page that contains the latest commit. This one should be straightforward: `strcmp()` the file names and compare the binlog offsets as unsigned numbers.

@Thirunarayanan can implement these changes.

Reply:
Good catch.

XIDs can be sorted by sequence number. Function read_wsrep_xid_seqno() can be used to read the sequence number part. One thing which seems to be missing though is a byte order conversion for the seqno, now the seqno is always stored in host byte order. Byte order conversion should be implemented in couple of places:
* storage/innobase/trx/trx0rseg.cc:read_wsrep_xid_seqno(), 
* sql/wsrep_xid.cc:wsrep_xid_init()
* sql/wsrep_xid.cc:wsrep_xid_seqno()

There is also another thing which I noticed after submitting a patch: The UUID part of the XID may change if the server is taken out of the existing cluster and used to bootstrap a new cluster. In the original implementation this was not a problem as the XID was stored in the single place in sys header. However, now the XID resides in several rseg headers, and if they contain different UUIDs, it is impossible to decide which of those represents the current cluster. One solution for this problem would be keeping cached value of current cluster UUID and clear all wsrep XIDs in rseg headers if the newly written XID has different UUID. This should be safe since when the cluster UUID changes, the first event which triggers the trx_rseg_update_wsrep_checkpoint() call is always executed in isolation, i.e. no interleaving wsrep operations are allowed.

Reply:
Thank you for the clarification. Maybe you could contribute a revised patch that deals with this issue.

If this turns out to be too complicated, we can still revert to writing the WSREP XID to the TRX_SYS page only. Commit performance would be slightly degraded for Galera, but not for single-master replication commits that store the binlog position.